specs/0000-st-app/product-spec.md

```diff
@@ -0,0 +1,419 @@
+| `routes` | `Sequence[BaseRoute]` | `None` | Additional Starlette routes |
+| `middleware` | `Sequence[Middleware]` | `None` | Middleware stack for all requests |
+| `exception_handlers` | `Mapping` | `None` | Custom exception handlers |
+| `debug` | `bool` | `False` | Enable debug mode |
+
+### Return Value
+
+An ASGI-compatible application object that can be:
+
+- Auto-detected and run via `streamlit run app.py`
+- Run directly with any ASGI server (e.g., `uvicorn app:app`)
+
+### Behavior
+
+**App discovery:**
+
+When `streamlit run app.py` is invoked, Streamlit checks if the script contains an
+`st.App` instance (similar to [FastAPI CLI discovery](https://github.com/fastapi/fastapi-cli/blob/main/src/fastapi_cli/discover.py)) by checking for a `st.App` instance named `app`. If no `st.App` instance is found, Streamlit will run the script in traditional mode.
+
+**Lifespan execution order:**
+
+1. Streamlit Runtime starts (enables full `@st.cache_resource` support)
+2. User's lifespan runs (startup code)
+3. Server starts accepting connections
+4. ... application runs ...
+5. Server stops accepting connections
+6. User's lifespan runs (shutdown code)
+7. Streamlit Runtime stops
+
+If an error occurs during the lifespan, Streamlit will log the error and abort the startup.
+
+**Route protection:**
+
+Streamlit reserves certain route prefixes. User routes cannot override:
+
+| Route Prefix | Purpose |
+|--------------|---------|
+| `/_stcore/*` | Core Streamlit API (WebSocket, health, upload) |
+| `/media/*` | Media file serving |
+| `/component/*` | Custom component serving (v1) |
+
+Conflicting routes raise `ValueError` at startup.
+
+**Middleware ordering:**
+
+```
+Request → [User Middleware] → [Streamlit Middleware] → [Route Handler]
+Response ← [User Middleware] ← [Streamlit Middleware] ← [Route Handler]
+```
+
+User middleware wraps Streamlit's internal middleware, enabling:
+
+- Auth middleware to reject requests before reaching Streamlit
+- Logging middleware to see all requests/responses
+- Security headers to be added to all responses
+
+### How `st.App` Addresses Feature Requests
+
+#### Custom HTTP Routes
+
+> Enables: [#439](https://github.com/streamlit/streamlit/issues/439), [#9673](https://github.com/streamlit/streamlit/issues/9673), [#6195](https://github.com/streamlit/streamlit/issues/6195), [#9090](https://github.com/streamlit/streamlit/issues/9090)
+
+```python
+import streamlit as st
+from starlette.routing import Route, Mount
+from starlette.responses import JSONResponse, PlainTextResponse, FileResponse
+from starlette.staticfiles import StaticFiles
+
+# REST API endpoint (#439)
+async def api_data(request):
+    return JSONResponse({"data": [1, 2, 3]})
+
+# robots.txt at root (#9673)
+async def robots_txt(request):
+    return PlainTextResponse("User-agent: *\nDisallow: /admin/")
+
+# Serve generated site folder (#6195)
+generated_site = StaticFiles(directory="./generated_docs", html=True)
+
+app = st.App(
```

Comment:
Small clarification: Do I need to do `app = st.App(...)` for the auto-detection to work, or can I also simply to `st.App(...)` without assigning it to a variable?